app/code/Magento/Catalog/Model/Product/SalabilityChecker.php

```diff
@@ -54,4 +63,23 @@ public function isSalable($productId, $storeId = null): bool
 
         return $product->isSalable();
     }
+    
+    /**
+     * Check if product is visible on frontend.
+     *
+     * @param int|string $productId
+     * @param int|null $storeId
+     * @return bool
+     */
+    public function isProductVisible($productId, $storeId = null): bool
+    {
+        if ($storeId === null) {
+            $storeId = $this->storeManager->getStore()->getId();
+        }
+        /** @var \Magento\Catalog\Model\Product $product */
+        $product = $this->productRepository->getById($productId, false, $storeId);
```

Comment:
There actually might be exception, please wrap it to try-catch block and return false if we weren’t able to find such product

Reply:
Done